packages/block-library/src/block/edit.js

```diff
@@ -1,180 +1,161 @@
-						onEdit={ this.startEditing }
-						onChangeTitle={ this.setTitle }
-						onSave={ this.save }
-						onCancel={ this.stopEditing }
-					/>
-				) }
-				{ ! isSelected && ! isEditing && <ReusableBlockIndicator title={ reusableBlock.title } /> }
-			</Fragment>
+			<RegistryProvider value={ this.registry }>
+				<EditorProvider
+					// Force a rerender when reusableBlock is updated, so that changes made
+					// externally appear in the nested editor.
+					key={ reusableBlockInstanceId }
+					postType="wp_block"
+					inheritContext
+					settings={ settingsWithLock }
+					post={ reusableBlock }
+				>
+					<ReusableBlockSelection
+						isReusableBlockSelected={ isSelected }
+						onBlockSelection={ setIsSelected }
+					>
+						{ list }
+						{ ( isSelected || isEditing ) && (
+							<ReusableBlockEditPanel
+								isEditing={ isEditing }
+								isSaving={ isSaving && ! reusableBlock.isTemporary }
+								onEdit={ this.startEdit }
+								onCancel={ this.cancelEdit }
+								onSave={ this.saveEdit }
+							/>
+						) }
+						{ ! isSelected && ! isEditing && <ReusableBlockIndicator /> }
+					</ReusableBlockSelection>
+				</EditorProvider>
+			</RegistryProvider>
 		);
 	}
 }
 
 export default compose( [
 	withSelect( ( select, ownProps ) => {
-		const {
-			getReusableBlock,
-			isFetchingReusableBlock,
-			isSavingReusableBlock,
-			getBlock,
-		} = select( 'core/editor' );
 		const { ref } = ownProps.attributes;
-		const reusableBlock = getReusableBlock( ref );
+		if ( ! Number.isFinite( ref ) ) {
+			return;
+		}
 
+		const { getEntityRecord } = select( 'core' );
 		return {
-			reusableBlock,
-			isFetching: isFetchingReusableBlock( ref ),
-			isSaving: isSavingReusableBlock( ref ),
-			block: reusableBlock ? getBlock( reusableBlock.clientId ) : null,
+			reusableBlock: getEntityRecord( 'postType', 'wp_block', ref ),
+			settings: select( 'core/editor' ).getEditorSettings(),
 		};
 	} ),
 	withDispatch( ( dispatch, ownProps ) => {
-		const {
-			fetchReusableBlocks,
-			updateBlockAttributes,
-			updateReusableBlockTitle,
-			saveReusableBlock,
-		} = dispatch( 'core/editor' );
-		const { ref } = ownProps.attributes;
-
+		const { selectBlock, receiveReusableBlocks } = dispatch( 'core/editor' );
+		const { receiveEntityRecords } = dispatch( 'core' );
 		return {
-			fetchReusableBlock: partial( fetchReusableBlocks, ref ),
-			updateAttributes: updateBlockAttributes,
-			onUpdateTitle: partial( updateReusableBlockTitle, ref ),
-			onSave: partial( saveReusableBlock, ref ),
+			setIsSelected() {
+				selectBlock( ownProps.clientId );
+			},
+			onSave( reusableBlock, parsedBlock ) {
+				// Update the editor's store when the reusable block is changed. This
+				// ensures other instances of the same reusable block are updated.
+				receiveReusableBlocks( [ { reusableBlock, parsedBlock } ] );
+				receiveEntityRecords( 'postType', 'wp_block', reusableBlock );
```

Comment:
We should be able to remove the `receiveReusableBlocks` dispatch here once we clean up how reusable blocks are fetched and stored.